sys/kern/kern_cheri_revoke.c

```diff
@@ -2,16 +2,605 @@
+	 */
+	if (entryst == CHERI_REVOKE_ST_NONE) {
+		epoch++;
+
+		KASSERT((myst == CHERI_REVOKE_ST_INITING) ||
+			(myst == CHERI_REVOKE_ST_CLOSING),
+			("Bad myst when finishing"));
+		entryst = CHERI_REVOKE_ST_INITED;
+
+		if (myst == CHERI_REVOKE_ST_CLOSING) {
+close_already_inited:	/* (entryst == CHERI_REVOKE_ST_INITED) above */
+
+			/* Walk the VM */
+			res = vm_cheri_revoke_pass(&vmcrc, 0);
+		}
+	}
+
+	/* OK, that's that.  Where do we stand now? */
+	if ((res == KERN_SUCCESS) &&
+	    (myst == CHERI_REVOKE_ST_CLOSING)) {
+
+#ifdef DIAGNOSTIC
+		vm_cheri_assert_consistent_clg(&vm->vm_map);
+#endif
+
+		/* Signal the end of this revocation epoch */
+		epoch++;
+		crepochs.dequeue = epoch;
+		vm_cheri_revoke_publish_epochs(info_page, &crepochs);
+		entryst = CHERI_REVOKE_ST_NONE;
+	}
+
+	vm_map_lock(vmm);
+	cheri_revoke_st_set(&vmm->vm_cheri_revoke_st, epoch, entryst);
+#ifdef CHERI_CAPREVOKE_STATS
+	if (flags & CHERI_REVOKE_TAKE_STATS) {
+		sx_xlock(&vmm->vm_cheri_revoke_stats_sx);
+		crst = *(struct cheri_revoke_stats*)&vmm->vm_cheri_revoke_stats;
+		crstp = &crst;
+		bzero(&vmm->vm_cheri_revoke_stats,
+		    sizeof(vmm->vm_cheri_revoke_stats));
+		sx_xunlock(&vmm->vm_cheri_revoke_stats_sx);
+	} else {
+		crstp = (struct cheri_revoke_stats*)&vmm->vm_cheri_revoke_stats;
+	}
+#endif
+	vm_map_unlock(vmm);
+
+	/* Broadcast here: some sleepers may be able to take the fast out */
+	cv_broadcast(&vmm->vm_cheri_revoke_cv);
+
+	vm_cheri_revoke_cookie_rele(&vmcrc);
+	vmspace_free(vm);
+
+	return (cheri_revoke_fini(crsi, vm_mmap_to_errno(res), crstp,
+	    &crepochs));
+}
+
+static int
+kern_cheri_revoke_get_shadow(struct thread *td, int flags,
+    void * __capability arena, void * __capability * __capability shadow)
+{
+	int arena_perms, error;
+	void * __capability cres;
+	vm_offset_t base, size;
+	int sel = flags & CHERI_REVOKE_SHADOW_SPACE_MASK;
+
+	if (!SV_CURPROC_FLAG(SV_CHERI)) {
+		return (ENOSYS);
+	}
+
+	switch (sel) {
+	case CHERI_REVOKE_SHADOW_NOVMMAP:
+
+		if (cheri_gettag(arena) == 0)
+			return (EINVAL);
+
+		arena_perms = cheri_getperm(arena);
+
+		if ((arena_perms & CHERI_PERM_SW_VMEM) == 0)
+			return (EPERM);
+
+		base = cheri_getbase(arena);
+		size = cheri_getlen(arena);
+
+		cres = vm_cheri_revoke_shadow_cap(curproc->p_sysent,
+			sel, base, size, arena_perms);
+
+		break;
+
+	case CHERI_REVOKE_SHADOW_OTYPE:
+	    {
+		int reqperms;
+
+		if (cheri_gettag(arena) == 0)
+			return (EINVAL);
+
+		/* XXX Require all of VMMAP, SEAL, and UNSEAL permissions? */
+		reqperms = CHERI_PERM_SEAL | CHERI_PERM_UNSEAL |
+		    CHERI_PERM_SW_VMEM;
+		arena_perms = cheri_getperm(arena);
+		if ((arena_perms & reqperms) != reqperms)
+			return (EPERM);
+
+		base = cheri_getbase(arena);
+		size = cheri_getlen(arena);
+
+		cres = vm_cheri_revoke_shadow_cap(curproc->p_sysent,
+			sel, base, size, 0);
+
+		break;
+	    }
+	case CHERI_REVOKE_SHADOW_INFO_STRUCT:
+	case CHERI_REVOKE_SHADOW_NOVMMAP_ENTIRE: // XXX
```

Comment:
Can we get rid of `NOVMMAP_ENTIRE`?

Reply:
Not yet. We're still using mrs rather than updating snmalloc. Hopefully for the next release? At a minimum, I'd like to make it possible to disable along with access to other, overly broad roots as a hardening measure.